EIPS/eip-erc721-marketplace-extension.md

```diff
@@ -0,0 +1,323 @@
+  * token transfer, such as signature-based, or token burning.
+  * 
+  * @param fromAddress_ : previous owner of the token
+  * @param toAddress_   : new owner of the token
+  * @param tokenId_     : identifier of the token being transferred
+  * 
+  * Emits a {Transfer} event.
+  */
+  function _transfer( address fromAddress_, address toAddress_, uint256 tokenId_ ) internal override {
+    if ( _isForSale[ tokenId_ ] ) {
+      _removeListing( tokenId_ );
+    }
+    super._transfer( fromAddress_, toAddress_, tokenId_ );
+  }
+
+  // PUBLIC
+  function listItem( uint256 tokenId, uint256 price, address to ) external {
+  /**
+  * @notice Create or update a listing for `tokenId_`.
+  * 
+  * Note: Setting `buyer_` to the NULL address will create a public listing.
+  * 
+  * @param tokenId_ : identifier of the token being listed
+  * @param price_   : the sale price of the token being listed
+  * @param buyer_   : optional address the token is being sold too
+  */
+  function listItem( uint256 tokenId_, uint256 price_, address buyer_ ) external {
+    address _tokenOwner_ = ownerOf( tokenId_ );
+    require( _tokenOwner_ == ownerOf( tokenId_ ), "Not token owner" );
+
+    _createListing( tokenId_, price_, _tokenOwner_, buyer_ );
+  }
+
+  /**
+  * @notice Removes the listing for `tokenId_`.
+  * 
+  * @param tokenId_ : identifier of the token being listed
+  */
+  function delistItem( uint256 tokenId_ ) external {
+    address _tokenOwner_ = ownerOf( tokenId_ );
+    require( _tokenOwner_ == ownerOf( tokenId_ ), "Not token owner" );
+
+    require( _isForSale[ tokenId_ ], "Invalid listing" );
+
+    _removeListing( _index_ );
+    emit Delisted( tokenId_ );
+  }
+
+  /**
+  * @notice Purchases the listed token `tokenId_`.
+  * 
+  * @param tokenId_ : identifier of the token being purchased
+  */
+  function buyItem( uint256 tokenId_ ) external payable {
+    require( _isForSale[ tokenId_ ], "Invalid listing" );
+    require(
+      msg.sender == _listings[ _listingIndex[ tokenId_ ] ].to ||
+      _listings[ _listingIndex[ tokenId_ ] ].to == address( 0 ),
+      "Invalid sale address"
+    )
+    require( msg.value == _listings[ _listingIndex[ tokenId_ ] ].price, "Incorrect price" );
+
+    address _tokenOwner_ = ownerOf( tokenId_ );
+    _tranfer( _tokenOwner_, msg.sender, tokenId_ );
+    emit Purchased( tokenId_, _tokenOwner_, msg.sender, _listing_.price );
+
+    // Handle royalties
+    ( address _royaltyRecipient_, uint256 _royalties_ ) = royaltyInfo( tokenId_, msg.value );
+    _processEthPayment( _royalties_, _royaltyRecipient_ );
+
+    uint256 _payment_ = msg.value - _royalties_;
+    _processEthPayment( _payment_, _tokenOwner_ );
+  }
+
+  // VIEW
+  /**
+  * @notice Returns the list of all existing listings.
+  * 
+  * @return the list of all existing listings
+  */
+  function getAllListings() external view returns ( Listing[] memory ) {
+    return _listings;
+  }
```

Comment:
This function doesn't seem to work. `solidity` doesn't seem to support returning arrays?

Reply:
I'll have to check that, I haven't had the time to do unit tests on the code suggested yet. Thanks for pointing it.

Reply:
We updated the code for reference
https://github.com/5660-eth/ERC721YY/blob/main/contract/ERC721YY.sol